wandb/sdk/launch/builder/build.py

```diff
@@ -499,3 +510,93 @@ def _create_docker_build_ctx(
 def join(split_command: List[str]) -> str:
     """Return a shell-escaped string from *split_command*."""
     return " ".join(shlex.quote(arg) for arg in split_command)
+
+
+def construct_builder_args(
+    launch_config: Optional[Dict] = None,
+    build_config: Optional[Dict] = None,
+) -> Tuple[Dict[str, Any], Dict[str, Any], Dict[str, Any]]:
+    registry_config = None
+    given_docker_args = {}
+    if launch_config is not None:
+        given_docker_args = launch_config.get("docker", {}).get("args", {})
+        build_config = launch_config.get("build")
+        registry_config = launch_config.get("registry")
+
+    default_launch_config = None
+    if os.path.exists(os.path.expanduser(LAUNCH_CONFIG_FILE)):
+        with open(os.path.expanduser(LAUNCH_CONFIG_FILE)) as f:
+            default_launch_config = yaml.safe_load(f)
+
+    build_config, registry_config = resolve_build_and_registry_config(
+        default_launch_config, build_config, registry_config
+    )
+
+    return given_docker_args, build_config, registry_config
+
+
+def build_image_with_builder(
+    builder: AbstractBuilder,
+    launch_project: LaunchProject,
+    repository: Optional[Any],
+    entry_point: EntryPoint,
+    docker_args: dict,
+) -> Optional[str]:
+    """
+    Helper for testing and logging
+    """
+    wandb.termlog(f"{LOG_PREFIX}Building docker image from uri source")
+    image_uri: Optional[str] = builder.build_image(
+        launch_project,
+        repository,
+        entry_point,
+        docker_args,
+    )
+    return image_uri
+
+
+def build_image_from_project(
+    launch_project: LaunchProject,
+    api: Api,
+    launch_config: Optional[Dict] = None,
+    default_builder_type: Optional[str] = "docker",
+) -> str:
+    """
+    Accepts a reference to the Api class and a pre-computed launch_spec
+    object, with an optional launch_config to set git-things like repository
+    which is used in naming the output docker image, and build_type defaulting
+    to docker (but could be used to build kube resource jobs w/ "kaniko")
+
+    updates launch_project with the newly created docker image uri and
+    returns the uri
+    """
+    assert launch_project.uri, "To build an image on queue a URI must be set."
+
+    # TODO(gst): launch_project.repository support w/ CLI arg (pref over config)
+    repository = None
+    if launch_config:
+        repository = launch_config.get("repository")
+    docker_args, builder_config, _ = construct_builder_args(launch_config)
```

Comment:
The third argument returned here is the registry config, which contains a key `url` this value is the repository value you want to use.
Lets use that if the repository is not set by the CLI arg or launch config.